File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointList/mountpoint/EditFragment.java
package ru.nsu.bobrofon.easysshfs.mountpointList.mountpoint;


import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.TextView;

import com.topjohnwu.superuser.Shell;

import java.util.Locale;

import ru.nsu.bobrofon.easysshfs.DrawerStatus;
import ru.nsu.bobrofon.easysshfs.EasySSHFSActivity;
import ru.nsu.bobrofon.easysshfs.R;
import ru.nsu.bobrofon.easysshfs.mountpointList.MountPointsList;

import static android.app.Activity.RESULT_OK;

public class EditFragment extends Fragment {
	private static final String MOUNT_POINT_ID = "MOUNT_POINT_ID";
	private static final int PICKDIR_REQUEST_CODE = 1;
	private static final int PICK_IDENTITY_FILE_CODE = 2;

	private int mMountPointId;
	private DrawerStatus mDrawerStatus;
	private MountPoint mSelf;

	private TextView mName;
	private CheckBox mAuto;
	private TextView mUsername;
	private TextView mHost;
	private TextView mPort;
	private TextView mPassword;
	private CheckBox mStorePassword;
	private TextView mIdentityFile;
	private TextView mRemotePath;
	private TextView mLocalPath;
	private CheckBox mForcePermissions;
	private TextView mOptions;

	public void setDrawerStatus(final DrawerStatus drawerStatus) {
		mDrawerStatus = drawerStatus;
	}

	public static EditFragment newInstance(final int mountPointId) {
		EditFragment fragment = new EditFragment();
		Bundle args = new Bundle();
		args.putInt(MOUNT_POINT_ID, mountPointId);
		fragment.setArguments(args);
		return fragment;
	}

	public EditFragment() {
		// Required empty public constructor
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getArguments() != null) {
			mMountPointId = getArguments().getInt(MOUNT_POINT_ID);
		}
	}

	public static String sdcard()
	{
		if (isMultiUserEnvironment())
		{
			return "/data/media/0";
		} else
		{
			return "/mnt/sdcard";
		}
	}

	public static boolean isMultiUserEnvironment()
	{
		return android.os.Build.VERSION.SDK_INT >= 17;
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		super.onCreateView(inflater, container, savedInstanceState);
		setHasOptionsMenu(true);

		// Inflate the layout for this fragment
		final View selfView = inflater.inflate(R.layout.fragment_edit, container, false);

		final MountPointsList worker
			= MountPointsList.getIntent(getActivity());

		if (worker.getMountPoints().size() > mMountPointId) {
			mSelf = worker.getMountPoints().get(mMountPointId);
		}
		else {
			mSelf = new MountPoint();
			mSelf.setRootDir(getActivity().getFilesDir().getPath());
			mSelf.setLocalPath(sdcard() + "/mnt");
		}

		mName = selfView.findViewById(R.id.mount_point_name);
		mAuto = selfView.findViewById(R.id.automount);
		mUsername = selfView.findViewById(R.id.username);
		mHost = selfView.findViewById(R.id.host);
		mPort = selfView.findViewById(R.id.port);
		mPassword = selfView.findViewById(R.id.password);
		mStorePassword = selfView.findViewById(R.id.store_password);
		mRemotePath = selfView.findViewById(R.id.remote_path);
		mLocalPath = selfView.findViewById(R.id.local_path);
		mForcePermissions = selfView.findViewById(R.id.force_permissions);
		mOptions = selfView.findViewById(R.id.sshfs_options);
		Button selectLocalDir = selfView.findViewById(R.id.select_dir);
		selectLocalDir.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
					selectLocalDir();
				}
			}
		});
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
			selectLocalDir.setEnabled(false);
		}
		mIdentityFile = selfView.findViewById(R.id.identity_file);
		final Button selectIdentityFile = selfView.findViewById(R.id.identity_file_select);
		selectIdentityFile.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				selectIdentityFile();
			}
		});

		mName.setText(mSelf.getPointName());
		mAuto.setChecked(mSelf.getAutoMount());
		mUsername.setText(mSelf.getUserName());
		mHost.setText(mSelf.getHost());
		mPort.setText(String.format(Locale.getDefault(), "%d", mSelf.getPort()));
		mPassword.setText(mSelf.getPassword());
		mStorePassword.setChecked(mSelf.getStorePassword());
		mRemotePath.setText(mSelf.getRemotePath());
		mLocalPath.setText(mSelf.getLocalPath());
		mForcePermissions.setChecked(mSelf.getForcePermissions());
		mOptions.setText(mSelf.getOptions());
		mIdentityFile.setText(mSelf.getIdentityFile());

		return selfView;
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);
		if (mDrawerStatus == null || !mDrawerStatus.isDrawerOpen()) {
			inflater.inflate(R.menu.edit, menu);
		}
	}

	private void grabMountPoint(final MountPoint mountPoint) {
		mountPoint.setPointName(mName.getText().toString());
		mountPoint.setAutoMount(mAuto.isChecked());
		mountPoint.setUserName(mUsername.getText().toString());
		mountPoint.setHost(mHost.getText().toString());
		mountPoint.setPort(mPort.getText().toString());
		mountPoint.setPassword(mPassword.getText().toString());
		mountPoint.setStorePassword(mStorePassword.isChecked());
		mountPoint.setIdentityFile(mIdentityFile.getText().toString());
		mountPoint.setRemotePath(mRemotePath.getText().toString());
		mountPoint.setLocalPath(mLocalPath.getText().toString());
		mountPoint.setForcePermissions(mForcePermissions.isChecked());
		mountPoint.setOptions(mOptions.getText().toString());
		mountPoint.setRootDir(getActivity().getFilesDir().getPath());
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		int id = item.getItemId();

		if (id == R.id.action_save) {
			grabMountPoint(mSelf);

			final MountPointsList worker = MountPointsList.getIntent(getActivity());

			if (!worker.getMountPoints().contains(mSelf)) {
				worker.getMountPoints().add(mSelf);
			}
			worker.save(getActivity());
			showToast("saved");

			return true;
		}
		else if (id == R.id.action_delete) {
			final MountPointsList worker
				= MountPointsList.getIntent(getActivity());
			worker.getMountPoints().remove(mSelf);
			worker.save(getActivity());
			showToast("deleted");

			return true;
		}
		else if (id == R.id.action_mount) {
			MountPoint mountPoint = new MountPoint();
			grabMountPoint(mountPoint);
			mountPoint.mount(true, getContext(), getShell());
		}
		else if (id == R.id.action_umount) {
			MountPoint mountPoint = new MountPoint();
			grabMountPoint(mountPoint);
			mountPoint.umount(true, getContext(), getShell());
		}

		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onAttach(Activity activity) {
		super.onAttach(activity);
		((EasySSHFSActivity) activity).onSectionAttached(R.string.mount_point_title);
	}

	private void showToast(final String message) {
		EasySSHFSActivity.showToast(message, getContext());
	}

	private Shell getShell() {
		return ((EasySSHFSActivity) getActivity()).getShell();
	}

	@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
	private void selectLocalDir() {
		Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
		startActivityForResult(intent, PICKDIR_REQUEST_CODE);
	}

	private void selectIdentityFile() {
		Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
		intent.addCategory(Intent.CATEGORY_OPENABLE);
		intent.setType("*/*");
		startActivityForResult(Intent.createChooser(intent, "Select IdentityFile"),
			PICK_IDENTITY_FILE_CODE);
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		switch (requestCode) {
			case PICKDIR_REQUEST_CODE:
				if (resultCode == RESULT_OK) {
					final Uri localUrl = data.getData();
					mLocalPath.setText(FileUtil.getFullPathFromTreeUri(localUrl, getContext()));
				}
				break;
			case PICK_IDENTITY_FILE_CODE:
				if (resultCode == RESULT_OK) {
					final Uri localUrl = data.getData();
					final String path = FileUtil.getPath(getContext(), localUrl);
					replaceIdentityFile(path);
				}
				break;
			default:
				super.onActivityResult(requestCode, resultCode, data);
		}
	}

	private void replaceIdentityFile(final String path) {
		if (path == null || path.isEmpty()) {
			return;
		}
		String options = mOptions.getText().toString();
		options = options.replaceAll(",?IdentityFile=[^,]*,?", ",")
			.replaceAll(",$|^,", "");
		mOptions.setText(options);
		mIdentityFile.setText(path);
	}
}

File: app/src/main/java/ru/nsu/bobrofon/easysshfs/mountpointlist/mountpoint/EditFragment.kt
// SPDX-License-Identifier: MIT
package ru.nsu.bobrofon.easysshfs.mountpointlist.mountpoint

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.*
import androidx.activity.result.contract.ActivityResultContracts.GetContent
import androidx.activity.result.contract.ActivityResultContracts.OpenDocumentTree
import androidx.annotation.RequiresApi
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle

import ru.nsu.bobrofon.easysshfs.EasySSHFSActivity
import ru.nsu.bobrofon.easysshfs.EasySSHFSFragment
import ru.nsu.bobrofon.easysshfs.R
import ru.nsu.bobrofon.easysshfs.databinding.FragmentEditBinding
import ru.nsu.bobrofon.easysshfs.mountpointlist.MountPointsList

class EditFragment : EasySSHFSFragment() {

    private var mountPointId: Int = 0
    private lateinit var mountPointsList: MountPointsList
    private lateinit var self: MountPoint

    private var _binding: FragmentEditBinding? = null
    private val binding get() = _binding!!
    private val name get() = binding.mountPointName
    private val auto get() = binding.automount
    private val username get() = binding.username
    private val hostname get() = binding.host
    private val port get() = binding.port
    private val password get() = binding.password
    private val storePassword get() = binding.storePassword
    private val remotePath get() = binding.remotePath
    private val localPath get() = binding.localPath
    private val forcePermissions get() = binding.forcePermissions
    private val options get() = binding.sshfsOptions
    private val identityFile get() = binding.identityFile
    private val selectLocalDir get() = binding.selectDir
    private val selectIdentityFile get() = binding.identityFileSelect

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mountPointId = arguments?.getInt(MOUNT_POINT_ID)
            ?: throw IllegalStateException("$MOUNT_POINT_ID argument is required")
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        super.onCreateView(inflater, container, savedInstanceState)

        val context = requireContext()

        mountPointsList = MountPointsList.instance(context)

        self = if (mountPointsList.mountPoints.size > mountPointId) {
            mountPointsList.mountPoints[mountPointId]
        } else {
            MountPoint().apply {
                rootDir = context.filesDir.path
                localPath = sdcard() + "/mnt"
            }
        }

        _binding = FragmentEditBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initLocalDirSelector(selectLocalDir)
        selectIdentityFile.setOnClickListener { selectIdentityFile() }

        name.setText(self.visiblePointName)
        auto.isChecked = self.autoMount
        username.setText(self.userName)
        hostname.setText(self.host)
        port.setText(self.port.toString())
        password.setText(self.password)
        storePassword.isChecked = self.storePassword
        remotePath.setText(self.remotePath)
        localPath.setText(self.localPath)
        forcePermissions.isChecked = self.forcePermissions
        options.setText(self.options)
        identityFile.setText(self.identityFile)

        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(menuProvider, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    private fun initLocalDirSelector(selector: View) {
        selector.setOnClickListener {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                selectLocalDir()
            }
        }
        selector.isEnabled = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP
    }

    private fun grabMountPoint(mountPoint: MountPoint) {
        val context = requireContext()

        mountPoint.pointName = name.text.toString()
        mountPoint.autoMount = auto.isChecked
        mountPoint.userName = username.text.toString()
        mountPoint.host = hostname.text.toString()
        mountPoint.setPort(port.text.toString())
        mountPoint.password = password.text.toString()
        mountPoint.storePassword = storePassword.isChecked
        mountPoint.identityFile = identityFile.text.toString()
        mountPoint.remotePath = remotePath.text.toString()
        mountPoint.localPath = localPath.text.toString()
        mountPoint.forcePermissions = forcePermissions.isChecked
        mountPoint.options = options.text.toString()
        mountPoint.rootDir = context.filesDir.path
    }

    private val menuProvider = object : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            if (!drawerStatus.isDrawerOpen) {
                menuInflater.inflate(R.menu.edit, menu)
            }
        }

        override fun onMenuItemSelected(menuItem: MenuItem): Boolean =
            when (menuItem.itemId) {
                R.id.action_save -> {
                    saveAction()
                    true
                }

                R.id.action_delete -> {
                    deleteAction()
                    true
                }

                R.id.action_mount -> {
                    mountAction()
                    true
                }

                R.id.action_umount -> {
                    umountAction()
                    true
                }

                else -> false
            }
    }

    private fun saveAction() {
        val context = requireContext()

        grabMountPoint(self)

        if (!mountPointsList.mountPoints.contains(self)) {
            mountPointsList.mountPoints.add(self)
        }
        mountPointsList.save(context)
        showToast("saved")
    }

    private fun deleteAction() {
        val context = requireContext()

        mountPointsList.mountPoints.remove(self)
        mountPointsList.save(context)
        showToast("deleted")
    }

    private fun mountAction() {
        val shell = shell!!

        val mountPoint = MountPoint()
        grabMountPoint(mountPoint)
        mountPoint.mount(shell, context)
    }

    private fun umountAction() {
        val shell = shell!!

        val mountPoint = MountPoint()
        grabMountPoint(mountPoint)
        mountPoint.umount(shell, context)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        appActivity?.onSectionAttached(R.string.mount_point_title)
    }

    private fun showToast(message: String) {
        EasySSHFSActivity.showToast(message, context)
    }

    private val localDirPicker =
        registerForActivityResult(OpenDocumentTree()) { uri: Uri? ->
            uri?.let { setLocalPath(it) }
        }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun selectLocalDir() {
        localDirPicker.launch(/* starting location */ null)
    }

    private val identityFilePicker =
        registerForActivityResult(object : GetContent() {
            override fun createIntent(context: Context, input: String): Intent {
                val intent = super.createIntent(context, input)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                return Intent.createChooser(intent, "Select IdentityFile")
            }
        }) { uri: Uri? ->
            uri?.let { setIdentityFile(it) }
        }

    private fun selectIdentityFile() {
        identityFilePicker.launch("*/*")
    }

    private fun setLocalPath(uri: Uri) {
        val context = requireContext()

        val path = FileUtil.getFullPathFromTreeUri(uri, context)
        // Most of the time users select some directory, they are actually trying to select a path
        // on one of the bind-mounted runtime directories (like "/mnt/runtime/[default|write|read|...]/"),
        // because they just want to access the remote filesystem from most of their Android applications.
        // So let's replace the path they selected with the path in "/mnt/runtime/default" directory.
        // This is not necessary that a user would want us to do, but he can edit this path manually
        // in case we are wrong.
        val runtimeStoragePath = toUnderlyingDefaultStorage(path)
        localPath.setText(runtimeStoragePath)
    }

    private fun setIdentityFile(uri: Uri) {
        val context = requireContext()

        val path = FileUtil.getPath(uri, context) ?: return
        replaceIdentityFile(path)
    }

    private fun replaceIdentityFile(path: String) {
        var optionString = options.text.toString()
        optionString = optionString.replace(",?IdentityFile=[^,]*,?".toRegex(), ",")
            .replace(",$|^,".toRegex(), "")
        options.setText(optionString)
        identityFile.setText(path)
    }

    companion object {
        private const val MOUNT_POINT_ID = "MOUNT_POINT_ID"

        fun newInstance(id: Int): EditFragment {
            val fragment = EditFragment()
            val args = Bundle()
            args.putInt(MOUNT_POINT_ID, id)
            fragment.arguments = args
            return fragment
        }

        private fun sdcard(): String = when {
            // TODO(bobrofon): substitute an actual active user id instead of hardcoding '0'
            hasRuntimePermissions -> {
                "/mnt/runtime/default/emulated/0"
            }

            isMultiUserEnvironment -> {
                "/data/media/0"
            }

            else -> {
                "/mnt/sdcard"
            }
        }

        private fun toUnderlyingDefaultStorage(storagePath: String): String {
            return if (hasRuntimePermissions) {
                storagePath.replaceFirst(Regex("^/storage/"), "/mnt/runtime/default/")
            } else {
                storagePath
            }
        }

        private val isMultiUserEnvironment: Boolean
            get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 // Android 4.2

        private val hasRuntimePermissions: Boolean
            get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M // Android 6.0
    }
} // Required empty public constructor
